WIP: Add multi-factor authentication plug-able modules

awarecan opened the pull request:
## Description:
Add multi-factor authentication support for new auth system.

### Auth module
Multi-factor auth modules has been added to auth/modules folder. 

`totp` is Time-based One Time Password module which compatible with Google Authenticator and Authy (not test yet). 

`insecure_example` is a Example designed for demo and unit testing purpose, should not be used in any production system.

Multi-facot auth module can be used mixed-match with auth providers. After normal auth provider validate, if there are auth modules enabled for the specific user, the login flow will direct to auth module input form. Auth module use separated storage, can be hardening it in future.

**Related issue (if applicable):** fixes #<home-assistant issue number goes here>

**Pull request in [developers.home-assistant.io](https://github.com/home-assistant/developers.home-assistant) with documentation (if applicable):** home-assistant/developers.home-assistant#52

## Example entry for `configuration.yaml` (if applicable):
```yaml
# configuration.yaml
homeassistant:
  # Auth Provider
  auth_providers:
    - type: homeassistant
  auth_mfa_modules:
    - type: totp

auth:

```

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Local tests pass with `tox`. **Your PR cannot be merged unless tests pass**

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [developers.home-assistant.io](https://github.com/home-assistant/developers.home-assistant)

If the code communicates with devices, web services, or third-party tools:
  - [x] New dependencies have been added to the `REQUIREMENTS` variable ([example][ex-requir]).
  - [x] New dependencies are only imported inside functions that use them ([example][ex-import]).
  - [x] New or updated dependencies have been added to `requirements_all.txt` by running `script/gen_requirements_all.py`.

If the code does not interact with devices:
  - [x] Tests have been added to verify that the new code works.

[ex-requir]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L14
[ex-import]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L54

nielstron committed:
Added bandpass filter

Allows values in a given range

nielstron committed:
Add test for new band_pass filter

nielstron committed:
Reordered attribute order

nielstron committed:
Fixed test

nielstron committed:
Removed default values and fixed description in sensor.filter

nielstron committed:
Renamed to range filter

nielstron committed:
Fix indent

nielstron committed:
Changes for new FilterState construct

nielstron committed:
Fix wrong comparison

nielstron committed:
Fix doctring newline and handle ha.state string-being

nielstron committed:
Fixed wrong bound assignment on values below the lower bound

nielstron committed:
Removal of windows size and precision for range filter

nielstron committed:
Fixed Rangefilter constructor call

lbschenkel committed:
deconz: fix light.turn_off with transition (#15222)

When light.turn_off is invoked with a transition, the following payload was
sent to deCONZ via PUT to /light/N/state:

{ "bri": 0, "transitiontime": transition }

However, on recent versions of deCONZ (latest is 2.05.31 at the time of
writing) this does not turn off the light, just sets it to minimum brightness
level (brightness is clamped to minimum level the light supports without
turning it off).

This commit makes the code send this payload instead:

{ "on": false, "transitiontime": transition }

This works as intended and the light does transition to the 'off' state.
This change was tested with Philips Hue colored lights, IKEA colored lights
and IKEA white spectrum lights: they were all able to be turned off
successfully with the new payload, and none of them could be turned off with
the old payload.

awarecan committed:
Add multi factor authentication support

Only include base class and insecure_example

pepsonEL committed:
Add new RTS device (#15116)

* Add new RTS device

Add new RTS Somfy device as cover-ExteriorVenetianBlindRTSComponent

* add next device

add next device

carlchan committed:
Add additional parameters to NUT UPS sensor (#15066)

* Update nut.py

Added input.frequency and a number of output parameters.

* Update nut.py

Fixed formatting issues
Added "devices" fields

* Separated "device.description" line to two lines.

* Update nut.py

Removed device.* sensors

fabaff committed:
Fix typos (#15233)

cdce8p committed:
Use async syntax for cover platforms (#15230)

scop committed:
Upgrade pytest to 3.6.2 (#15241)

Yevgenium committed:
Add precipitations to Openweathermap daily forecast mode (#15240)

* Add precipitations to daily forecast mode

* Remove line breaks

lbschenkel committed:
deconz: proper fix light.turn_off with transition (#15227)

Previous commit d4f7dfa successfully fixed the bug in which lights
would not turn off if a transition was specified, however if 'bri' is not
present in the payload of the PUT request set to deCONZ, then any
'transitiontime' ends up being ignored. This commit addresses the
unintended side effect by reintroducing 'bri', resulting in the following
payload:

{ "on": false, "bri": 0, "transitiontime": ... }

dthulke committed:
allow extra slot values in intents (#15246)

scop committed:
Lint cleanups (#15243)

* Remove some unused imports

* Fix a flake8 E271

balloob committed:
Bump frontend to 20180701.0

fabaff committed:
Upgrade sqlalchemy to 1.2.9 (#15250)

fabaff committed:
Upgrade WazeRouteCalculator to 0.6 (#15251)

SConaway committed:
Fix typo in Docker files (#15256)

andrey-git committed:
Switch to own packaged version of pylgnetcast (#15042)

## Description:

Switch to own packaged version of pylgnetcast

Request to make a pypi package didn't get any response: https://github.com/wokar/pylgnetcast/issues/1

**Related issue (if applicable):** #7069

kstaniek committed:
Added setting cover tilt position in scene (#15255)

## Description:
This feature adds possibly of setting tilt_position in scene for covers.

**Related issue (if applicable):** fixes #<home-assistant issue number goes here>

**Pull request in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io) with documentation (if applicable):** home-assistant/home-assistant.github.io#<home-assistant.github.io PR number goes here>

## Example entry for `configuration.yaml` (if applicable):
```yaml
scene:
  - name: Close Cover Tilt
    entities:
      cover.c_office_north:
        tilt_position: 0

  - name: Open Cover Tilt
    entities:
      cover.c_office_north:
        tilt_position: 100
```

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Local tests pass with `tox`. **Your PR cannot be merged unless tests pass**

If user exposed functionality or configuration variables are added/changed:
  - [ ] Documentation added/updated in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io)

If the code communicates with devices, web services, or third-party tools:
  - [ ] New dependencies have been added to the `REQUIREMENTS` variable ([example][ex-requir]).
  - [ ] New dependencies are only imported inside functions that use them ([example][ex-import]).
  - [ ] New or updated dependencies have been added to `requirements_all.txt` by running `script/gen_requirements_all.py`.
  - [ ] New files were added to `.coveragerc`.

If the code does not interact with devices:
  - [ ] Tests have been added to verify that the new code works.

[ex-requir]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L14
[ex-import]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L54

scop committed:
Pass tox posargs to pylint (#15226)

arbreng committed:
Fix Roomba exception (#15262)

* Fix Roomba exception

* Switch to single quotes

balloob committed:
Update frontend to 20180702.0

Commit:
Added support for Duke Energy smart meters (#15165)

* Added support for Duke Energy smart meters

* Fixed hound

* Added function docstring

* Moved strings to constants, implemented unique_id, and cleaned up setup.

* Added doc string.

* Fixed review issues.

* Updated pydukenergy to 0.0.6 and set update interval to 2 hours

* Updated requirements_all

dgomes committed:
Update image_processing async (#15082)

* scan() -> async_job

* added async_scan

balloob committed:
Bump frontend to 20180702.1

shkertt committed:
Fix python-miio 0.4 compatibility of the xiaomi miio device tracker (#15244)

fabaff committed:
Upgrade keyring to 13.1.0 (#15268)

Kane610 committed:
deCONZ - new sensor attribute 'on' and new sensor GenericFlag (#15247)

* New sensor attribute 'on'
* New sensor GenericFlag

nielstron committed:
Remove math.inf as bounds

stenius committed:
expose climate current temperature in prometeus metrics (#15232)

* expose climate current temperature in prometeus metrics

* import ATTR_CURRENT_TEMPERATURE from climate instead of const

* remove duplicated ATTR_CURRENT_TEMPERATURE from const

* fix ATTR_CURRENT_TEMPERATURE import

pepsonEL committed:
New device to support option MY in somfy (#15272)

New device to support option MY in somfy

fabaff committed:
Add isort configuration (#15278)

nielstron committed:
Remove default none from filter sensor

fabaff committed:
Add support for new API (fixes #14911) (#15279)

andrey-git committed:
Switch to own packaged version of suds-passworddigest (#15261)

balloob committed:
Bump frontend to 20180703.0

balloob committed:
Disable the calendar panel (#15282)

balloob committed:
Bump frontend to 20180703.1

bachya committed:
Update Tile platform to be async (#15073)

* Updated

* Updated requirements

* Added expired session handling

* Changes

* Member-requested changes

* Bump to 2.0.2

* Bumping requirements

* Better exception handling and tidying

* Move asyncio stuff to HASS built-ins

* Revising re-initi

* Hound

* Hound

cdce8p committed:
Fix yeelight light brightness integer (#15290)

dgomes committed:
Added Push Camera (#15151)

* Added push camera

* add camera.push

* Address comments and add tests

* auff auff

* trip time made no sense

* travis lint

* Mock dependency

* hound

* long line

* long line

* better mocking

* remove blank image

* no more need to mock dependency

* remove import

* cleanup

* no longer needed

* unused constant

* address @pvizeli review

* add force_update

* Revert "add force_update"

This reverts commit e203785ea8232722effeec4fc70358190ec9284a.

* rename parameter

tchellomello committed:
Added support to HTTPS URLs on SynologyDSM (#15270)

* Added support to HTTPS URLs on SynologyDSM

* Bumped python-synology to 0.1.1

* Makes lint happy

* Added attribution to Synology and fixed 3rd library version

* Fixed requirements_all.txt

* Makes SynologyDSM defaults to 5001 using SSL

balloob committed:
Add system generated users (#15291)

* Add system generated users

* Fix typing

balloob committed:
Bump frontend to 20180704.0

bachya committed:
Fix exception when parts of Pollen.com can't be reached (#15296)

Fix exception when parts of Pollen.com can't be reached

lukiffer committed:
Add additional sensors for Arlo Baby camera (#15074)

* Add additional sensors for Arlo Baby camera

* Fix linter errors

* Fix linter error

* Add tests for Arlo sensors

* Fix linter errors

* Bump pyarlo dependency to 0.1.9

* Remove unnecessary AttributeError except

* Fix module reference error in py35

* Fix test

* Address PR review concerns

* Convert to standalone pytest methods

* Fix linter errors

* Fix linter errors

* Fix linter errors

* Fix test

* Remove redundant check, fix async test

* Fix linter error

* Added check for total_cameras sensor, added additional attribute tests

* Add missing docstring

worm-ee committed:
Add HomematicIP Cloud Config Flow and Entries loading (#14861)

* Add HomematicIP Cloud to config flow

* Inititial trial for config_flow

* Integrations text files

* Load and write config_flow and init homematicip_cloud

* Split into dedicated files

* Ceanup of text messages

* Working config_flow

* Move imports inside a function

* Enable laoding even no accesspoints are defined

* Revert unnecassary changes in CONFIG_SCHEMA

* Better error handling

* fix flask8

* Migration to async for token generation

* A few fixes

* Simplify config_flow

* Bump version to 9.6 with renamed package

* Requirements file

* First fixes after review

* Implement async_step_import

* Cleanup for Config Flow

* First tests for homematicip_cloud setup

* Remove config_flow tests

* Really remove all things

* Fix comment

* Update picture

* Add support for async_setup_entry to switch and climate platform

* Update path of the config_flow picture

* Refactoring for better tesability

* Further tests implemented

* Move 3th party lib inside function

* Fix lint

* Update requirments_test_all.txt file

* UPdate of requirments_test_all.txt did not work

* Furder cleanup in websocket connection

* Remove a test for the hap

* Revert "Remove a test for the hap"

This reverts commit 968d58cba108e0f371022c7ab540374aa2ab13f4.

* First tests implemented for config_flow

* Fix lint

* Rework of client registration process

* Implemented tests for config_flow 100% coverage

* Cleanup

* Cleanup comments and code

* Try to fix import problem

* Add homematicip to the test env requirements

awarecan commented:
For other auth provider, such as `legacy_api_password`, you need to manually edit `~/.homeassistant/.storage/auth_module.totp` file (the username for legacy_api_password is `homeassistant`)

To manually generate `ota_secret`, you can execute following python script
```python
import pyotp
print(pyotp.random_base32())
```

fabaff committed:
Upgrade youtube_dl to 2018.07.04 (#15323)

fabaff committed:
Upgrade keyring to 13.2.0 (#15322)

scop committed:
Upgrade pytest to 3.6.3 (#15332)

quazzie committed:
Add original message as dialogflow_query parameter (#15304)

So you can access for example sessionId as {{ dialogflow_query.sessionId }} in intent templates.

andrey-git committed:
Add python 3.7 to travis and tox (#14523)

* Add python 3.7 to travis and tox

* Use pyyaml from github

* Don't version constraints

* Fix version tag

* Change to new pyyaml release

* Python 3.7 requires xenial

* Fix namespace detection

* Use correct RegEx type

* Update pexpect to 4.6

* Use correct validation for dictionaries

* Disable Py37 incompatible packages

* Upgrade all pexpect to 4.6

* Add explicit None as default param

balloob commented on homeassistant/auth.py:
```diff
@@ -77,6 +243,35 @@ def name(self):
         """Return the name of the auth provider."""
         return self.config.get(CONF_NAME, self.DEFAULT_TITLE)
 
+    async def load_modules(self, moudle_configs):
```
typo

balloob commented on homeassistant/auth_providers/homeassistant.py:
```diff
@@ -44,9 +47,12 @@ def __init__(self, hass):
+                'users': [],
             }
 
+        if 'users' not in data:
```
When can this happen?

awarecan replied:
Data file was corrupted? Cannot recall why I added this statement, can be removed if you feel redundant.

balloob replied:
It's redundant. If storage doesn't exist, we get `None` back and set the default above. If we save, the default was the foundation.

balloob commented:
It's usually wise to discuss these things before creating big PRs like this. That way there is no work wasted if we have different views on the architecture.

balloob commented on homeassistant/auth_providers/homeassistant.py:
```diff
@@ -118,16 +124,19 @@ class HassAuthProvider(auth.AuthProvider):
+            await self.hass.async_add_executor_job(
+                data.validate_login, username, password)
+        finally:
+            await data.async_save()
```
Why do we need to always save?

awarecan replied:
legacy code, should be removed. It was used in v1 implementation when auth module save data in  auth provider storage.

awarecan commented:
I thought we had the discussion around plugable auth module in #15225

balloob committed:
Bump frontend to 20180708.0

sjabby committed:
Frontend: Allow overriding default url when added to home screen (#15368)

Frontend: Allow overriding default url when added to home screen

worm-ee committed:
Add HomematicIP Cloud light power consumption and energie attributes (#15343)

* Add power consumption and energie attributes

* Fix lint

* Change attribute name and include kwh

dgomes committed:
fix camera.push API overwrite (#15334)

* fix camera.push API overwrite

* dont search in the component dictionary, but in hour own

* remove error message

* hound

NoUsername committed:
Add support for multi-channel enocean switches (D2-01-12 profile) (#14548)

starkillerOG committed:
Add sound mode support (#14910)

* Add sound mode support

* continuation line indent

* indentation

* indentation

* Remove option to configure sound_mode_dict

* Sound mode support

- removed the sound_mode_raw propertie because it was not used, (still available through self._sound_mode_raw (as device attribute for automations and diagnostics)

* Detect sound mode support from device

Removed the config option to indicate if sound mode is supported.
Added detection if sound mode is supported from the receiver itself.
Pushed denonavr library to V.0.7.4

* Pushed denonavr to v.0.7.4

iliketoprogram14 committed:
Fixed issue 15340. alexa/smart_home module can now skip properties that aren't supported in the current state, eg lowerSetpoint in Heat mode or targetSetpoint in Eco mode for Nest devices. (#15352)

balloob commented:
That's true, but that's only 1 part of this. Starting code review now

balloob commented on homeassistant/auth_providers/modules/insecure_example.py:
```diff
@@ -0,0 +1,66 @@
+"""Example auth module."""
```
This should not live under auth provider as it has nothing to do with it? 

I suggest we restructure our auth into

- `homeassistant/auth/__init__.py`
- `homeassistant/auth/providers/homeassistant.py`
- `homeassistant/auth/modules/totp.py`

I also not sure if I like `modules`, it's too generic. What about `two_factor` ?

balloob commented on homeassistant/auth_providers/modules/insecure_example.py:
```diff
@@ -0,0 +1,66 @@
+}, extra=vol.PREVENT_EXTRA)
+
+STORAGE_VERSION = 1
+STORAGE_KEY = 'auth_module.insecure_example'
```
I think that we should store these on the user objects in a data object that is stored/restored when user is saved/restored.

balloob commented on homeassistant/auth.py:
```diff
@@ -121,6 +317,77 @@ def async_create_credentials(self, data):
         return {}
 
 
+class LoginFlow(data_entry_flow.FlowHandler):
```
This is the wrong architecture. 

I think that we should implement it as a FlowHandler that is not extended by the login flows, but instead wraps one.

It should take the login flow it needs to wrap. Once that flow returns a create entry, it should hold that value, look up the user, see if any two factor has been registered, instantiate those and let them handle the next step.

You can use `__getattr__` to intercept the methods on the LoginFlow and decide where to redirect the call.

balloob replied:
Hmm , maybe I do like this architecture. I just don't like how it iterates through the modules and mutates them.

balloob commented on homeassistant/auth.py:
```diff
@@ -121,6 +317,77 @@ def async_create_credentials(self, data):
+                    if not errors and result:
+                        return await self.async_finish(result)
+                else:
+                    session_id = await auth_module.\
```
We shouldn't create sessions, just instantiate and store on this instance.

balloob commented on homeassistant/auth.py:
```diff
@@ -121,6 +317,77 @@ def async_create_credentials(self, data):
+
+    async def async_finish(self, username):
+        """Handle the pass of login flow."""
+        if self._auth_modules:
```
Use a guard clause

```python
if not X:
    return Y

…
```

balloob commented on homeassistant/auth.py:
```diff
@@ -121,6 +317,77 @@ def async_create_credentials(self, data):
+    async def async_finish(self, username):
+        """Handle the pass of login flow."""
+        if self._auth_modules:
+            _, auth_module = self._auth_modules.popitem(False)
```
This is weird, instead, once finish called, we should check if user has any 2fa configured (stored on user object?) and then create a local list of 2fa to ask which we can mutate.

balloob replied:
A user can have multiple 2FA active, it only needs to pass 1 to log in.

awarecan replied:
I see your point, so the flow should like

admin create user -> user login -> user enable/setup one of many 2fa from a list
user select auth_provider -> auth_provider validate -> user select which 2fa he/she want to use -> ask 2fa challenge -> 2fa module validate -> user logged in

balloob replied:
Yep, and we can do it all within a login flow without needed a frontend change

balloob commented on homeassistant/auth.py:
```diff
@@ -121,6 +317,77 @@ def async_create_credentials(self, data):
+        if self._auth_modules:
+            _, auth_module = self._auth_modules.popitem(False)
+
+            step_id = 'auth_module_' + auth_module.type
```
string concatenation should be done using `.format`

balloob commented on homeassistant/auth.py:
```diff
@@ -46,6 +66,151 @@ def generate_secret(entropy: int = 32) -> str:
+        await self.async_save()
+        return data
+
+    async def async_cleanup_session(self):
```
Session expiration is up to the LoginFlow to enforce.

balloob commented on homeassistant/auth.py:
```diff
@@ -23,18 +26,35 @@
     vol.Optional(CONF_NAME): str,
     # Specify ID if you have two auth providers for same type.
     vol.Optional(CONF_ID): str,
+    vol.Optional(CONF_MODULES):
```
Modules should not be nested under an auth provider. Instead, users should be able to enable 2fa for their account.

awarecan replied:
So the configuration should like

```yaml
# configuration.yml
homeassistant:
  auth_providers:
    - type: a_provder
    - type: b_provider
  auth_2fa_modules:
    - type: totp
    - type: hotp_sms
```

user should be able to enable one of many module in his setting page
user should be able to select use which module after auth_provider validate

That is the sequence used in my Cisco VPN client configured with okta integration.

fabaff committed:
Efergy (#15380)

* Update format

* Use string formatting

awarecan commented on homeassistant/auth.py:
```diff
@@ -46,6 +66,151 @@ def generate_secret(entropy: int = 32) -> str:
+    """Raised when we encounter invalid authentication."""
+
+
+class SessionStore:
```
Yes, it was inherited from previous version of implement. I agreed that is over-design, we should not have session at all, flow_id can be used as a session.

balloob committed:
Use IndieAuth for client ID (#15369)

* Use IndieAuth for client ID

* Lint

* Lint & Fix tests

* Allow local IP addresses

* Update comment

balloob committed:
Bump frontend to 20180709.0

scop committed:
Add httplib2 to h.c.google requirements (#15385)

scop committed:
Remove some unneeded pylint import-error disables (#15386)

danielperna84 committed:
Update ha-philipsjs to 0.0.5 (#15378)

* Update requirements_all.txt

* Update philips_js.py

hanzoh committed:
Add new voices to Amazon Polly (#15320)

ludeeus committed:
Add Cloudflare DNS component. (#15388)

* Add Cloudflare DNS component

* Removed man

* Update .coveragerc

* Update cloudflare.py

* Update cloudflare.py

* Changed records to be required

* Fix typos, update order and other minor changes

homeassistant commented:
Hello @awarecan,

When attempting to inspect the commits of your pull request for CLA signature status among all authors we encountered commit(s) which were not linked to a GitHub account, thus not allowing us to determine their status(es).

The commits that are missing a linked GitHub account are the following:

- [`6f9a98b674fcefd9e2d5165915da4db18529ee30`](https:&#x2F;&#x2F;github.com&#x2F;home-assistant&#x2F;home-assistant&#x2F;pull&#x2F;15335&#x2F;commits&#x2F;6f9a98b674fcefd9e2d5165915da4db18529ee30) - This commit has something that looks like an email address (ec2-user@ip-172-31-28-229.us-west-2.compute.internal). Maybe try linking that to GitHub?.

Unfortunately, **we are unable to accept this pull request until this situation is corrected.**

Here are your options:

1. If you had an email address set for the commit that simply wasn't linked to your GitHub account you can link that email now and it will retroactively apply to your commits. The simplest way to do this is to click the link to one of the above commits and look for a blue question mark in a blue circle in the top left. Hovering over that bubble will show you what email address you used. Clicking on that button will take you to your email address settings on GitHub. Just add the email address on that page and you're all set. GitHub has more information about this option [in their help center](https://help.github.com/articles/setting-your-email-in-git/#commits-on-github-arent-linking-to-my-account).

2. If you didn't use an email address at all, it was an invalid email, or it's one you can't link to your GitHub, you will need to change the authorship information of the commit and your global Git settings so this doesn't happen again going forward. GitHub provides some great instructions on how to change your authorship information [in their help center](https://help.github.com/articles/setting-your-email-in-git/).
   * If you only made a single commit you should be able to run
      ```
      git commit --amend --author="Author Name <email@address.com>"
      ```
      (substituting Author Name and email@address.com for your actual information) to set the authorship information.
   * If you made more than one commit and the commit with the missing authorship information is not the most recent one you have two options:
       1. You can re-create all commits missing authorship information. This is going to be the easiest solution for developers that aren't extremely confident in their Git and command line skills.
       2. You can use [this script](https://help.github.com/articles/changing-author-info/) that GitHub provides to rewrite history. **Please note:** this should be used only if you are very confident in your abilities and understand its impacts.
   * Whichever method you choose, I will come by to re-check the pull request once you push the fixes to this branch.

We apologize for this inconvenience, especially since it usually bites new contributors to Home Assistant. We hope you understand the need for us to protect ourselves and the great community we all have built legally. The best thing to come out of this is that you only need to fix this once and it benefits the entire Home Assistant and GitHub community.

Thanks, I look forward to checking this PR again soon! :heart:

robmarkcole committed:
Add Facebox teach service (#14998)

* Adds service

* Address pylint

* Update facebox.py

* patch tests

* Update facebox.py

* Update test_facebox.py

* Update facebox.py

* Update facebox.py

* Update facebox.py

* Update test_facebox.py

* Update test_facebox.py

* Update facebox.py

* Update facebox.py

* Update facebox.py

* Update facebox.py

* Adds total_matched_faces

* Update test_facebox.py

* Update facebox.py

* Update test_facebox.py

* Update test_facebox.py

* Remove fixtures

Removes the fixtures which were causing `setup` to fail, replace with `@patch`

* Fix teach service test and lint issues

awarecan commented:
Half-way in the refactoring
- [x] changed the directory structure
- [x] remove session
- [ ] move auth module config out of auth provider
- [ ] change login flow to allow user select auth module

balloob committed:
Expire auth code after 10 minutes (#15381)

balloob committed:
Bump frontend to 20180710.0

glpatcern committed:
Improve NetAtmo sensors update logic (#14866)

* Added a "last update" sensor that could be used by automations + cosmetic changes

* Improved the update logic of sensor data

The platform is now continuously adjusting the refresh interval
in order to synchronize with the expected next update from the
NetAtmo cloud. This significantly improves reaction time of
automations while keeping the refresh time to the recommended
value (10 minutes).

* Linting

* Incorporated the advanced Throttle class to support adaptive
throttling, as opposed to integrating it in the core framework.

Following code review, it was suggested to implement the
specialised Throttle class in this platform instead of making a
change in the general util package. Except that the required change
(about 4 LoC) is part of the only relevant piece of code of that
class, therefore this commit includes a full copy of the Throttle
class from homeassistant.util, plus the extra feature to support
adaptive throttling.

* Cosmetic changes on the introduced "last updated" sensor

* Alternate implementation for the adaptive throttling

Ensure the updates from the cloud are throttled and adapted to the
last update time provided by NetAtmo, without using the Throttle
decorator. Similar logic and similar usage of a lock to protect
the execution of the remote update.

* Linting

ludeeus committed:
removed unused return (#15402)

amelchio committed:
Fix confused brightness of xiaomi_aqara gateway light (#15314)

pschmitt committed:
Fix liveboxplaytv empty channel list (#15404)

balloob committed:
Fix credentials lookup (#15409)

tchellomello committed:
Change Ring binary_sensor frequency polling to avoid rate limit exceeded errors (#15414)

awarecan commented on homeassistant/scripts/auth.py:
```diff
@@ -32,54 +35,114 @@ def run(args):
+            {'username': args.username})
+        user = await auth_manager.async_get_or_create_user(credential)
+        secret = await auth_manager.async_enable_user_mfa(user, 'totp')
+        # FIXME need to wait until AuthStore save finish
```
@balloob I cannot find a way to wait AuthStore.async_save() actually done, since it will add a task delay 1 sec start.

balloob replied:
Try `await hass.async_stop()`

balloob replied:
Or `await asyncio.sleep(1.1)` 😉 (it's a script so fine with me)

awarecan replied:
`await hass.async_stop()` works, leave two INFO log entries on console (stop event), but I think it is fine.

awarecan commented:
Code is ready for review, document still need update

huangyupeng committed:
Add Tuya component and switch support (#15399)

* support for tuya platform

* support tuya platform

* lint fix

* change dependency

* add tuya platform support

* remove tuya platform except switch. fix code as required

* fix the code as review required

* fix as required

* fix a mistake

balloob commented on homeassistant/auth/__init__.py:
```diff
@@ -46,6 +57,66 @@ def generate_secret(entropy: int = 32) -> str:
+    """Raised when we encounter invalid authentication."""
+
+
+class AuthModule:
```
We should not call it Module. It's way too generic. Maybe TwoFactorModule.

Also, this class should be moved into that folder `auth/two_factor_module/__init__.py`

awarecan replied:
I have used MFA, so I will change AuthModule to MultiFactorAuthModule and change folder to auth/mfa_module

balloob commented on homeassistant/auth/__init__.py:
```diff
@@ -46,6 +57,66 @@ def generate_secret(entropy: int = 32) -> str:
+                      ))
+
+    @property
+    def id(self):  # pylint: disable=invalid-name
```
I don't think that we need to copy the id approach of providers. There can be only one.

awarecan replied:
They can have multiple instance, for example, I am planning to implement a mfa_module to use `hotp` algorithm, then send the code by one of configed HA notify service. So I will end up to have two module with same type, but have different config if I want to use email or SMS.

balloob commented on homeassistant/auth/__init__.py:
```diff
@@ -46,6 +57,66 @@ def generate_secret(entropy: int = 32) -> str:
+        return self.config[CONF_TYPE]
+
+    @property
+    def name(self):
```
We don't need name. The module should just pick a name.

balloob commented on homeassistant/auth/__init__.py:
```diff
@@ -186,13 +369,20 @@ class Credentials:
 
 
-async def load_auth_provider_module(hass, provider):
+async def load_module(hass, module_name, module_type):
```
This method too should be moved to `two_factor_module/__init__.py`

balloob commented on homeassistant/auth/__init__.py:
```diff
@@ -238,14 +428,35 @@ class Credentials:
     return manager
 
 
 async def _auth_provider_from_config(hass, store, config):
```
Should be moved to `auth/provider/__init__.py`

balloob commented on homeassistant/scripts/auth.py:
```diff
@@ -32,54 +35,115 @@ def run(args):
     data = hass_auth.Data(hass)
     loop.run_until_complete(data.async_load())
-    loop.run_until_complete(args.func(data, args))
+    loop.run_until_complete(args.func(hass.auth, data, args))
```
We should pass `hass`

danielperna84 committed:
Fix HomeMatic variables (#15417)

* Update __init__.py

* Update requirements_all.txt

mjg59 committed:
Ignore some HomeKit devices (#15316)

There are some devices that speak HomeKit that we shouldn't expose. Some
bridges (such as the Hue) provide reduced functionality over HomeKit and
have a functional native API, so should be ignored. We also shouldn't
offer to configure the built-in Home Assistant HomeKit bridge.

amelchio committed:
Make LimitlessLED color/temperature attributes mutually exclusive (#15298)

worm-ee committed:
Add HomematicIP alarm control panel (#15342)

* Add HomematicIP security zone

* Update access point tests

* Fix state if not armed and coments

* Add comment for the empty state_attributes

* Fix comment

* Fix spelling

awarecan committed:
Include request.path in legacy api password warning message (#15438)

balloob commented on homeassistant/auth/__init__.py:
```diff
@@ -186,36 +111,13 @@ class Credentials:
-async def auth_manager_from_config(hass, provider_configs):
+async def auth_manager_from_config(hass, provider_configs, module_configs):
     """Initialize an auth manager from config."""
+    # pylint: disable=protected-access
```
Instead of disabling the pylint, we should not have them be prefixed with a `_`

balloob commented on homeassistant/auth/__init__.py:
```diff
@@ -290,14 +197,40 @@ def support_legacy(self):
+    async def async_get_auth_module(self, module_id):
+        """Return an auth modules, None if not found."""
+        module = self._mfa_modules.get(module_id)
+        if module and not module.initialized:
```
We should drop initialization. That is up to the modules to do when they want.

balloob commented on homeassistant/auth/__init__.py:
```diff
@@ -341,8 +280,52 @@ def async_auth_providers(self):
+
+        await self._store.async_remove_credentials(credentials)
+
+    async def async_enable_user_mfa(self, user, mfa_module_id, **kwargs):
```
We shouldn't use `**kwargs`, if the module expects certain data, we should just store that in a dictionary and pass that on.

balloob commented on homeassistant/auth/__init__.py:
```diff
@@ -481,6 +459,42 @@ def __init__(self, hass):
+
+        local_user.mfa_modules.append(mfa_module_id)
+        await self.async_save()
+        return local_user
```
We should not return `local_user`

balloob commented on homeassistant/auth/__init__.py:
```diff
@@ -481,6 +459,42 @@ def __init__(self, hass):
+
+        local_user.mfa_modules.remove(mfa_module_id)
+        await self.async_save()
+        return local_user
```
We should not return `local_user`

balloob commented on homeassistant/auth/__init__.py:
```diff
@@ -481,6 +459,42 @@ def __init__(self, hass):
+
+    async def async_enable_user_mfa(self, user, mfa_module_id):
+        """Enable a mfa module for user."""
+        local_user = await self.async_get_user(user.id)
```
We should not look up the user. We should trust the passed in data.

balloob commented on homeassistant/auth/mfa_modules/__init__.py:
```diff
@@ -0,0 +1,132 @@
+
+    hass.data[DATA_REQS] = set()
+
+    req_success = await requirements.async_process_requirements(
```
Looks like we never add `module_name` to `processed`

awarecan replied:
Fixed

balloob commented on homeassistant/auth/mfa_modules/insecure_example.py:
```diff
@@ -0,0 +1,86 @@
+class InsecureExampleModule(MultiFactorAuthModule):
+    """Example auth module validate pin."""
+
+    DEFAULT_TITLE = 'Personal Identify Number'
```
Since this is the insecure example, it should have insecure in the title.

balloob commented on homeassistant/auth/providers/insecure_example.py:
```diff
@@ -16,22 +16,18 @@
 }, extra=vol.PREVENT_EXTRA)
 
 
-class InvalidAuthError(HomeAssistantError):
```
Let's restore this. There is no reason to share an exception class just because it has a similar name. This is an internal exception and should stay so.

balloob commented on homeassistant/auth/providers/legacy_api_password.py:
```diff
@@ -3,39 +3,36 @@
 LEGACY_USER = 'homeassistant'
 
 
-class InvalidAuthError(HomeAssistantError):
```
This is an internal error and should stay so. 

Also, changing this is scope creep…

balloob commented on homeassistant/scripts/auth.py:
```diff
@@ -32,54 +35,124 @@ def run(args):
         data.validate_login(args.username, args.password)
-        print("Auth valid")
-    except hass_auth.InvalidAuth:
+        if args.code:
```
We should use guard clauses.

```python
if not args.code:
    print(…)
    return
```

balloob commented on homeassistant/scripts/auth.py:
```diff
@@ -32,54 +35,124 @@ def run(args):
+                print("Auth invalid")
+        else:
+            print("Auth valid")
+    except auth.InvalidAuth:
```
What was raising this? And why? Please split the mfa part away from the valid auth part.

balloob commented:
This PR should be split in three parts.

 - Refactor auth 
 - Add mfa and just add insecure example to show how it works
 - Add totp mfa module

Right now it's not clear how the design completely works.

andrey-git committed:
Add python 3.8-dev to travis and tox (#15347)

* Add Python 3.8-dev tox tests.

* Allow failures on 3.8-dev

* Allow failures on 3.8-dev take2

* Only run on pushes to dev

balloob committed:
Reorg auth (#15443)

balloob commented:
I've merged a reorg of the auth in #15443 . I've used `git mv` so the changes should be minimal. I had little trouble rebasing #15420

balloob commented:
(make sure you squash all your commits before the rebase)

andrey-git committed:
Make typing checks more strict (#14429)

## Description:

Make typing checks more strict: add `--strict-optional` flag that forbids implicit None return type. This flag will become default in the next version of mypy (0.600)

Add `homeassistant/util/` to checked dirs.

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Local tests pass with `tox`. **Your PR cannot be merged unless tests pass**

scop committed:
upgrade-mypy (#14904)

* Upgrade mypy to 0.600

* Upgrade mypy to 0.610

* Typing improvements

* remove unneeded or

* remove merge artifact

* Update loader.py

balloob committed:
Fix comment formatting (#15447)

awarecan commented on homeassistant/auth/__init__.py:
```diff
@@ -71,14 +95,40 @@ def support_legacy(self):
         """Retrieve a user."""
         return await self._store.async_get_user(user_id)
 
+    async def async_get_user_by_credentials(self, credentials):
```
No create, only get

balloob commented:
I will merge my user management, that way it will slim down this PR even further, making it more visible what things are added to support MFA.

balloob committed:
User management (#15420)

* User management

* Lint

* Fix dict

* Reuse data instance

* OrderedDict all the way

balloob committed:
Bump frontend to 20180713.0

balloob commented on homeassistant/auth/mfa_modules/__init__.py:
```diff
@@ -0,0 +1,125 @@
+        raise NotImplementedError
+
+    async def async_depose_user(self, user_id):
+        """Setup auth module for user."""
```
stale comment

balloob commented on homeassistant/auth/mfa_modules/__init__.py:
```diff
@@ -0,0 +1,125 @@
+        raise NotImplementedError
+
+    async def async_validation(self, user_id, user_input):
+        """Return the validation result."""
```
what type of result, boolean ?

balloob commented on homeassistant/auth/mfa_modules/insecure_example.py:
```diff
@@ -0,0 +1,85 @@
+        self._users = config.get('users', [])
+
+    @property
+    def input_schema(self):
```
This should return a voluptuous schema.

balloob commented on homeassistant/auth/mfa_modules/insecure_example.py:
```diff
@@ -0,0 +1,85 @@
+        """Initialize the user data store."""
+        super().__init__(hass, config)
+        self._data = None
+        self._users = config.get('users', [])
```
No need for `get`, users is a required field.

balloob commented on homeassistant/auth/mfa_modules/insecure_example.py:
```diff
@@ -0,0 +1,85 @@
+}, extra=vol.PREVENT_EXTRA)
+
+STORAGE_VERSION = 1
+STORAGE_KEY = 'auth_module.insecure_example'
```
`mfa_modules.insecure_example`

balloob commented on homeassistant/auth/mfa_modules/insecure_example.py:
```diff
@@ -0,0 +1,85 @@
+        return schema
+
+    @property
+    def users(self):
```
Doesn't seem to be used?

awarecan replied:
I do have usage, but should not have.

balloob commented on homeassistant/auth/mfa_modules/insecure_example.py:
```diff
@@ -0,0 +1,85 @@
+            if user and user.get('user_id') == user_id:
+                # already setup, override
+                user['pin'] = pin
+                return pin
```
Why would we return the pin ?

awarecan replied:
The return needs for `totp`. I will remove it from `insecure_example`.

It may little bit weird, some modules need return, some not.

balloob commented on homeassistant/auth/mfa_modules/insecure_example.py:
```diff
@@ -0,0 +1,85 @@
+        if not data:
+            raise ValueError('Expect data parameter')
+
+        pin = data.get('pin')
```
This can't happen if part of input schema?

awarecan replied:
No, it is not, this is `setup_schema`, may not same as `input_schema` in general.

balloob commented on homeassistant/auth/mfa_modules/insecure_example.py:
```diff
@@ -0,0 +1,85 @@
+
+    async def async_validation(self, user_id, user_input):
+        """Return True if validation passed."""
+        if user_id is None or user_input is None:
```
How can this be none ?

awarecan replied:
This is core validation. I feel like I need guard it tight.

balloob commented on homeassistant/auth/providers/__init__.py:
```diff
@@ -137,3 +143,113 @@ def async_create_credentials(self, data):
+                errors['base'] = 'invalid_auth'
+
+        schema = OrderedDict()
+        schema['multi_factor_auth_module'] = vol.In(self._user.mfa_modules)
```
if `len(self._user.mfa_modules) == 1`, pick that automatically without promting the user to pick

awarecan replied:
It was considered in `async_finish`, will actually be called wailer than this method

balloob commented on homeassistant/auth/providers/__init__.py:
```diff
@@ -137,3 +143,113 @@ def async_create_credentials(self, data):
+            schema = OrderedDict()
+            schema['multi_factor_auth_module'] = vol.In(self._user.mfa_modules)
+
+            return self.async_show_form(
```
What? Don't do this. Instead, just `return await async_step_select_mfa_module()`

balloob commented on homeassistant/auth/providers/__init__.py:
```diff
@@ -137,3 +143,113 @@ def async_create_credentials(self, data):
+
+        return self.async_show_form(
+            step_id='mfa',
+            data_schema=vol.Schema(auth_module.input_schema),
```
auth module needs to generate their voluptuous schema.

balloob commented on homeassistant/auth/mfa_modules/insecure_example.py:
```diff
@@ -0,0 +1,85 @@
+        self._users.append({'user_id': user_id, 'pin': pin})
+        return pin
+
+    async def async_depose_user(self, user_id):
```
Let's name this `async_remove_user` ?

awarecan replied:
I fell that `async_remove_user` is more confusing, sounds like user been deleted, however I changed docstring.

balloob commented on homeassistant/auth/__init__.py:
```diff
@@ -83,6 +116,15 @@ def auth_providers(self):
+                if creds.id == credentials.id:
+                    return user
+
+        raise ValueError('Unable to find the user.')
```
Should return `None` as with the other get X methods.

awarecan committed:
Resolve code review comment

balloob commented on homeassistant/auth/__init__.py:
```diff
@@ -101,12 +143,7 @@ def auth_providers(self):
-                        return user
-
-            raise ValueError('Unable to find the user.')
+            return await self.async_get_user_by_credentials(credentials)
```
The behavior of this method has changed now.

balloob commented on homeassistant/auth/__init__.py:
```diff
@@ -83,6 +116,15 @@ def auth_providers(self):
         return await self._store.async_get_user(user_id)
 
+    async def async_get_user_by_credentials(self, credentials):
+        """Get a user by credential, raise ValueError if not found."""
```
Doc string is not correct.

balloob replied:
Generaly, if we're querying a value, `None` is an acceptable return value. If we're mutating, `ValueError` if the input value is incorrect.

balloob commented on homeassistant/auth/auth_store.py:
```diff
@@ -131,10 +149,10 @@ def __init__(self, hass):
             return
 
-        for user_dict in data['users']:
+        for user_dict in data.get('users', []):
```
We will always store users. This cannot happen.

balloob commented on homeassistant/auth/auth_store.py:
```diff
@@ -131,10 +149,10 @@ def __init__(self, hass):
             users[user_dict['id']] = models.User(**user_dict)
 
-        for cred_dict in data['credentials']:
+        for cred_dict in data.get('credentials', []):
```
Cannot happen either

balloob commented on homeassistant/auth/auth_store.py:
```diff
@@ -145,7 +163,7 @@ def __init__(self, hass):
         refresh_tokens = OrderedDict()
 
-        for rt_dict in data['refresh_tokens']:
+        for rt_dict in data.get('refresh_tokens', []):
```
Cannot happen either

balloob commented on homeassistant/auth/auth_store.py:
```diff
@@ -158,7 +176,7 @@ def __init__(self, hass):
             users[rt_dict['user_id']].refresh_tokens[token.token] = token
 
-        for ac_dict in data['access_tokens']:
+        for ac_dict in data.get('access_tokens', []):
```
Cannot happen either

balloob commented on homeassistant/auth/mfa_modules/insecure_example.py:
```diff
@@ -0,0 +1,84 @@
+    async def async_setup_user(self, user_id, data=None):
+        """Setup mfa module for user."""
+        try:
+            data = self.setup_schema(data)  # pylint: disable=not-callable
```
This shouldn't be needed. I am assuming that we use the same flow as providers:

Frontend renders setup_schema, and only allows passing data to this method if it's valid.

balloob commented on homeassistant/auth/mfa_modules/insecure_example.py:
```diff
@@ -0,0 +1,84 @@
+        schema['pin'] = str
+        return vol.Schema(schema)
+
+    async def async_setup_user(self, user_id, data=None):
```
I don't see how this can be `None`, at least it's a dict if no fields specified in setup schema

balloob commented on homeassistant/auth/mfa_modules/insecure_example.py:
```diff
@@ -0,0 +1,84 @@
+}, extra=vol.PREVENT_EXTRA)
+
+STORAGE_VERSION = 1
+STORAGE_KEY = 'mfa_modules.insecure_example'
```
This is not being used.

balloob commented on homeassistant/auth/mfa_modules/insecure_example.py:
```diff
@@ -0,0 +1,84 @@
+        """Validate async_setup_user input data."""
+        schema = OrderedDict()
+        schema['pin'] = str
+        return vol.Schema(schema)
```
So here it is getting kinda weird.

The way mfa modules work, is that we persist in the user which mfa are enabled, and then we ask the mfa module to show a form. I feel like this violates the single source of truth. We should probably not store which mfa modules are enabled for a user. Instead:

- MFA modules will get a method `async_is_enabled_for(user)`
- We can add a method to auth manager to gather enabled MFA modules for a user. It then asks each MFA module.

I think that since this module does not support setting up, it should return `None` for setup schema, which indicates to the frontend that this module cannot be setup.

What do you think of that?

balloob commented on homeassistant/auth/mfa_modules/insecure_example.py:
```diff
@@ -0,0 +1,84 @@
+    @property
+    def input_schema(self):
+        """Validate login flow input data."""
+        schema = OrderedDict()
```
No need for an ordered dict if there is just 1 field.

balloob commented on homeassistant/auth/providers/__init__.py:
```diff
@@ -137,3 +143,116 @@ def async_create_credentials(self, data):
+            # We need get user from username, so we need get credential first
+            # async_get_or_create_credentials would not save data, it is safe
+            # to be called here. It will be called later in end of workflow.
+            credentials = await self._auth_provider.\
```
Wait what's going on here?

Credentials are an opaque data format. Yes, our Home Assistant provider uses `{'username': username}` but that's just one auth provider. Credentials could also be a user ID in some external system etc.

`async_get_or_create_credentials` takes in the flow result.

balloob commented on homeassistant/auth/providers/__init__.py:
```diff
@@ -137,3 +143,116 @@ def async_create_credentials(self, data):
+        # new credential or no mfa_module enabled or passed mfa validate
+        return self.async_create_entry(
+            title=self._auth_provider.name,
+            data={'username': username}
```
This is wrong.

balloob commented on homeassistant/auth/providers/__init__.py:
```diff
@@ -137,3 +143,116 @@ def async_create_credentials(self, data):
+
+    async def async_finish(self, username, mfa_valid=False):
+        """Handle the pass of login flow."""
+        if not mfa_valid:
```
Use guard clauses.

```python
if mfa_valid:
    do something
    return

rest of code
```

balloob commented on homeassistant/scripts/auth.py:
```diff
@@ -25,6 +25,8 @@ def run(args):
     parser_list = subparsers.add_parser('list')
     parser_list.set_defaults(func=list_users)
+    parser_list.add_argument('-a', '--all', default=False,
+                             help="Show all users included system user")
```
This makes no sense. This is a script to manage the `homeassistant` auth provider. System generated users cannot have credentials linked and so will never be listed here

balloob commented on homeassistant/scripts/auth.py:
```diff
@@ -100,6 +114,7 @@ def run(args):
         print("Auth valid")
     except hass_auth.InvalidAuth:
         print("Auth invalid")
+        return
```
useless return?

balloob commented:
The current architecture is not correct. Makes incorrect assumptions about auth providers and how things relate to each other.

I have currently spend too much time on reviewing this PR and it has distracted me from my actual priorities. I won't be looking at this PR for the upcoming week.

awarecan commented:
The core issue here is: 

Current flow: after auth provider validate, we got flow result, starts from here, flow is finished, control back to auth manager, then it will use flow result to get or create credential, then get or create user, then login finish.

The challenge is, in mfa module, it needs user_id, we have do some hack here to try to get user by mfa module. That is the core issue.

If we want to do it in right way, we need to change data flow manager, to handle more flexible way, for example, I will suggest change FlowManager to

```python
class FlowManager:
    def __init__(self, hass, async_callback_flows: dict):
```

and the async_callback_flows should be something likes
```python
async_callback_flows = {
    'init': async_init_flow
    'get_user': async_get_user_flow
    'finish': async_finish_flow
}
```

awarecan commented:
Rebase seems mess up, close this PR for now
